### ISSUE TITLE
napari-lite

### ISSUE THREAD

**Issue by jni** [napari/napari#5940]
The main napari viewer provides a *lot* of functionality when one only wants to just look at an image. This leads to very slow initial launch times and import time. We should pare it down to the bare essentials and add things slowly and intentionally, leaving space for downstream apps, the app bundle, and plugins to provide specific functionality.

- [ ] Settings should not be loaded on import
- [ ] The lite viewer (which could one day be the default) should hide the layer list and most controls by default on launch
- [ ] The viewer would require fewer install-time dependencies
- [ ] instant launch time is *extremely important*
- [ ] menus, commands could all be skipped.
- [ ] ?

## Original issue

#4740

**Comment by AhmetCanSolak** [comment]
I love this @jni ! I agree with every item in your list, I would also suggest making python-console an optional dependency?

Also, in terms of file organization and packaging, how do you imagine this being implemented?

**Comment by jni** [comment]
LOL Long time no see @AhmetCanSolak!

Yes, making the console optional has been on my to-do since forever.

I don't have any specific ideas about the implementation at the moment though. Just riffing: we make a `napari.Canvas` object that has the viewport/scene and dimensions sliders, the napari Viewer uses that Canvas but the Canvas can live by itself.

**Comment by kephale** [comment]
Is this being connected to the fastplotlib explorations? e.g. maybe the fastplotlib prototype could just aim to satisfy the criteria of napari-lite

**Comment by kushalkolar** [comment]
Hello, our fastplotlib ImageWidget is kinda becoming a monster so maybe we could get together sometime sprint this out?  :smile: I'm quite bogged down until the holidays though. 

We just made fpl responsive to the `%gui qt` hook and implemented output contexts :D .

**Comment by kandarpksk** [comment]
During the October 2023 hackathon, a group of napari core devs brainstormed goals and non-goals for napari lite.
Here is a picture of the brainstorming board:
![napari lite brainstorm](https://github.com/napari/napari/assets/4083860/7942978e-4172-4336-ad6d-308031b6fc6b)

I have shared notes that I started writing up after the brainstorm separately with Draga (since they're unfinished).

**Comment by DragaDoncila** [comment]
This comment attempts to summarize the current state of the discussion napari-lite by aggregating commentary from the hackathon, the original issue, various discussions among core devs, etc.

## Motivation

1. The napari viewer has become cumbersome for rapid image viewing and iterative workflows. This is mainly due to slow install and slow start-up times.
2. The codebase has become increasingly complex and there are multiple refactors necessary (e.g. moving slicing entirely off layers, features refactor, deprecating action manager) before we can start thinking about some of the longer term features (e.g. napari on the web, serializability, swappable rendering backends).

## Goals
- Improved launch-time
    - `napari.imshow` should be almost instant
- Fewer install dependencies
    - how much can we really get rid of?
- Better separation of code base
    - napari-lite components should form the core of the full napari application
    - it should be easy for user/developers to extend the pared down viewer with different backends or introduce additional features from the app e.g. load layer controls only, or layerlist only
- Less magic
    - one current source of complexity is code designed to guess user expectations/intent e.g. guessing if an image is RGB, guessing which axis is the channel axis, guessing which reader you might want to use
    - napari-lite should avoid this code and provide explicit APIs for all inputs. There should be minimal autogeneration and napari should make no assumptions about user expectations

Target audience: Jupyter notebook user

## Desired Features
- canvas + canvas controls (dimension sliders & mouse/cursor info)
- colormapping/blending/overlays etc. programmatically only
- flexible for different backends
- code works identically in full viewer

### To Decide
- layerlist/multiple layers vs. single layer
    - layerlist programmatic only?
- layer controls?
    - lazily loaded?
    - can be added by users as a standard widget?
- painting/annotating
    - UI balloons **but** it's one of napari's key features
    - could it be a separate importable widget
    - a minimal set of controls?


## App-only Features
- settings
    - no persisted settings
    - preferences can still be set per instance?
- plugins
    - if user wants to use a reader/function/etc. they would import from the plugin package directly
    - there is no "plugin manager" associated with napari-lite
- menus/commands
    - everything is accessed through the viewer API
- keybindings
    - accessibility concerns?
    - we're already pretty inaccessible

## Implementation Considerations
- building "on top of" napari
    - napari-lite should not be built separately to napari
    - rather components should be separated and then integrated with the viewer one by one
- how to split out work into manageable PRs
- which refactors come first
- pre-requisites for starting this work?

**Comment by brisvag** [comment]
Great writeup, I generally agree with the state of the conversation. For me, big focus has to be on modularity/separation of the codebase and ease of contribution.

Some specific comments follow:

>     1. The napari viewer has become cumbersome for rapid image viewing and iterative workflows. This is mainly due to slow install and slow start-up times.

I think the goal here needs to be "use this refactoring to improve start times in general" and not "let's make napari-lite fast cause we can't improve napari-app".


> Target audience: Jupyter notebook user

Also devs who want to embed the viewer in other applications (e.g: @Czaki with PartSeg).

> ## Desired Features
> 
>     * canvas + canvas controls (dimension sliders & mouse/cursor info)
> 
>     * colormapping/blending/overlays etc. programmatically only


I would call these "must have". napari without these is barely useful :P

>     * code works identically in full viewer

Very important; a statement that came up often in discussions was "I want to be able to just change a line from `from napari_lite import Viewer` to `from napari import Viewer` and everything works.

> ### To Decide
> 
>     * layerlist/multiple layers vs. single layer
>       
>       * layerlist programmatic only?
> 
>     * layer controls?
>       
>       * lazily loaded?
>       * can be added by users as a standard widget?
> 
>     * painting/annotating
>       
>       * UI balloons **but** it's one of napari's key features
>       * could it be a separate importable widget
>       * a minimal set of controls?

I think these should be responsibilities of the napari app: one should be able to do something like `v = napari.Viewer(lite=True)` and skip all the heavy stuff, but then `viewer.unlite()` (or whatever xD) and get all the extra stuff. But `napari_lite` should know nothing of GUI layer controls and GUI layerlist. IMO, the only GUI should be sliders, and maybe the buttons at the bottom left, maybe. Or even expose only shortcuts for those actions...

>     * how to split out work into manageable PRs
> 
>     * which refactors come first

I think priority #1 is to decouple things that need decoupling (layers and slicing, canvas and vispy backend), which should help with reduce PR sizes as well.

>     * pre-requisites for starting this work?

Willpower and time :muscle:

**Comment by JoOkuma** [comment]
This would be really awesome!
Some of the items described above might cover it, but I would remove the thumbnails on `napari-lite`.
It slows the refresh quite a bit.

**Comment by psobolewskiPhD** [comment]
>  I would remove the thumbnails on `napari-lite`.
> It slows the refresh quite a bit.

@JoOkuma FYI: https://github.com/napari/napari/issues/6657

**Comment by jacopoabramo** [comment]
Not entirely sure if this is a sleeping issue but I'll try anyway: is there still some interest on this?

**Comment by DragaDoncila** [comment]
@jacopoabramo I guess you could say it is a napping issue? 😆 There is definitely still a lot of interest in this, and it's something the core developers are looking to tackle in earnest in the new year. We don't yet have a concrete plan of where we'll start and what we'll do, but we've had multiple discussions (mostly summarized above), about what it should be.

We would absolutely welcome any input on what you would envision it to be and any feedback you have that you think would be useful!

**Comment by jacopoabramo** [comment]
Sorry for the late reply!

My two cents are mostly related to what I'm trying to achieve, and specifically just have the image viewer. I want to strip it of the console, the plugin manager, and also the file readers and writers. I would like to just display raw numpy/xarray/dask data and also have access to all the layering mechanism within napari.

I don't know how exactly the project folder is structured, but are there some nested dependencies that would imply the impossibility of just creating an installation option with only the visualization features?

In the end napari started as a multi-dimensional visualization tool, the napari-lite version of it could be just sticking to this unopinionated version. Of course it would lose a lot of the functionalities that are useful for scientist, but again, this is for my specific use case that is unrelated to all of that.

**Comment by Czaki** [comment]
@jacopoabramo I think that you just want to use `QtViewer` instead of `Viewer`.  

```python
from napari.qt import QtViewer
from napari.components import ViewerModel

# create QtAplication instance here 

viewer_model = ViewerModel()
qt_viewer = QtViewer(viewer_model)
qt_viewer.show()
```

**Comment by jacopoabramo** [comment]
Ok, I did a very barebone benchmark:

```py
import time

start_time = time.time()

from napari.qt import QtViewer

print(f"QtViewer import completed: {time.time() - start_time:.4f} seconds")
start_time = time.time()
from napari.components import ViewerModel

print(f"ViewerModel import completed: {time.time() - start_time:.4f} seconds")

from qtpy.QtWidgets import QApplication
from qtpy.QtCore import QTimer

# Create QtApplication instance
app = QApplication([])

print(f"Application created: {time.time() - start_time:.4f} seconds")

# Create viewer model and QtViewer
model_creation_time = time.time()
viewer_model = ViewerModel()
qt_viewer = QtViewer(viewer_model)

print(f"Viewer model and QtViewer created: {time.time() - model_creation_time:.4f} seconds")

# Show the viewer window
show_time = time.time()
qt_viewer.show()
print(f"Viewer window shown: {time.time() - show_time:.4f} seconds")

# Exit the event loop immediately after starting
def exit_event_loop():
    print(f"Event loop started and exited: {time.time() - event_loop_time:.4f} seconds")
    app.quit()

# Use QTimer to exit the event loop immediately
QTimer.singleShot(0, exit_event_loop)

# Start the application event loop
event_loop_time = time.time()
app.exec_()
```

output:

```py
QtViewer import completed: 13.3013 seconds
ViewerModel import completed: 0.0000 seconds
Application created: 0.0308 seconds
Viewer model and QtViewer created: 1.0736 seconds
Viewer window shown: 0.2501 seconds
Event loop started and exited: 0.0006 seconds
```

What's inside QtViewer that's so heavy?

**Comment by jacopoabramo** [comment]
> @jacopoabramo I think that you just want to use `QtViewer` instead of `Viewer`.

I would also want to be able to not install all the dependencies tho. That and the slow startup time are my 2 major points that I wanted to understand better.

**Comment by psobolewskiPhD** [comment]
Very understandable. Just out of curiosity, here's my M1 Max, python 3.12, editable dev install from main:
```
QtViewer import completed: 2.4236 seconds
ViewerModel import completed: 0.0000 seconds
Application created: 0.1062 seconds
Viewer model and QtViewer created: 0.3442 seconds
Viewer window shown: 0.1743 seconds
Event loop started and exited: 0.0597 seconds
```

**Comment by jacopoabramo** [comment]
I realized that the env I was working on did not have a fresh version of napari so I updated everything and now:

First run:

```
QtViewer import completed: 14.3294 seconds
ViewerModel import completed: 0.0000 seconds
Application created: 0.0301 seconds
Viewer model and QtViewer created: 0.9078 seconds
Viewer window shown: 0.2299 seconds
Event loop started and exited: 0.0000 seconds

**Comment by jacopoabramo** [comment]
Once again: 

- Windows 11
- python 3.11

## Test 1:

> `conda create -n napari-pyqt python=3.11 napari pyqt`

First run execution time:

```
QtViewer import completed: 13.6404 seconds                                                                                                                                                                             
ViewerModel import completed: 0.0000 seconds                                                                                                                                                                           
Application created: 0.1302 seconds                                                                                                                                                                                    
Viewer model and QtViewer created: 0.9117 seconds                                                                                                                                                                      
Viewer window shown: 0.1982 seconds                                                                                                                                                                                    
Event loop started and exited: 0.0000 seconds
```

Second and consecutive runs execution time:

```
QtViewer import completed: 5.4041 seconds
ViewerModel import completed: 0.0000 seconds
Application created: 0.0131 seconds
Viewer model and QtViewer created: 0.8254 seconds
Viewer window shown: 0.1822 seconds
Event loop started and exited: 0.0000 seconds
```

## Test 2

> `conda create -n napari-pyqt python=3.11 napari pyside`

First run execution time:

```
QtViewer import completed: 13.3245 seconds
ViewerModel import completed: 0.0000 seconds
Application created: 0.1338 seconds
Viewer model and QtViewer created: 0.9260 seconds
Viewer window shown: 0.1771 seconds
Event loop started and exited: 0.0000 seconds
```

Second and consecutive runs execution time:

```
QtViewer import completed: 5.6112 seconds
ViewerModel import completed: 0.0000 seconds
Application created: 0.0132 seconds
Viewer model and QtViewer created: 0.8218 seconds
Viewer window shown: 0.1702 seconds
Event loop started and exited: 0.0000 seconds
```

Can someone with Windows confirm these timings with a quick test of the script?


```py
import time

start_time = time.time()

from napari.qt import QtViewer

print(f"QtViewer import completed: {time.time() - start_time:.4f} seconds")
start_time = time.time()
from napari.components import ViewerModel

print(f"ViewerModel import completed: {time.time() - start_time:.4f} seconds")

from qtpy.QtWidgets import QApplication
from qtpy.QtCore import QTimer

# Create QtApplication instance
app = QApplication([])

print(f"Application created: {time.time() - start_time:.4f} seconds")

# Create viewer model and QtViewer
model_creation_time = time.time()
viewer_model = ViewerModel()
qt_viewer = QtViewer(viewer_model)

print(f"Viewer model and QtViewer created: {time.time() - model_creation_time:.4f} seconds")

# Show the viewer window
show_time = time.time()
qt_viewer.show()
print(f"Viewer window shown: {time.time() - show_time:.4f} seconds")

# Exit the event loop immediately after starting
def exit_event_loop():
    print(f"Event loop started and exited: {time.time() - event_loop_time:.4f} seconds")
    app.quit()

# Use QTimer to exit the event loop immediately
QTimer.singleShot(0, exit_event_loop)

# Start the application event loop
event_loop_time = time.time()
app.exec_()
```

I can't decide if Windows is that bad or something weird is going on with my workstation

**Comment by Czaki** [comment]
Unfortunately, 5s for import is not bad. 
We import dask/pandas and a few other libraries that take significant time to import.

**Comment by jacopoabramo** [comment]
I understand that dask is necessary since it's what powers the lazy loading of multi dimensional data, but what about others?

By the way, I found an easier, more accurate (and harder to read) log file via `python -X importtime script.py`. It can be found attached.

[importtime_output.txt](https://github.com/user-attachments/files/17874122/importtime_output.txt)

**Comment by jacopoabramo** [comment]
Like, for my use case I don't need scipy and other packages like that

**Comment by psobolewskiPhD** [comment]
For the record, what I posted was with pyqt6.
Also, pandas we could probably replace? we don't use it per se. something like narwhals might work.

**Comment by Czaki** [comment]
We import dask only to setup cache parameters. 
We use pandas for properiets/features storage in Points/Shapes/Layers. 

I think that many of them may be optimzed/dellayed 

You may load yourt time import into 
https://kmichel.github.io/python-importtime-graph/ 

And see what happens. 

In general, someone just need to sit and check what happen.

**Comment by jacopoabramo** [comment]
Damn it's a mess.

> We use pandas for properiets/features storage in Points/Shapes/Layers.

What about [polars](https://github.com/pola-rs/polars)?

**Comment by andy-sweet** [comment]
> > We use pandas for properiets/features storage in Points/Shapes/Layers.
> 
> What about [polars](https://github.com/pola-rs/polars)?

Another related suggestion - the [DataFrame API standard](https://github.com/data-apis/dataframe-api) is further along now than it was when we introduced pandas (i.e. there is a draft we might be able to work off). That would allow napari to be dependency free for dataframe/table data, pushing the choice to the user/plugin-author when they need to support that kind of data.

**Comment by Czaki** [comment]
I don't know. People are more familiar with pandas. 

I do not work too much with this part of codebase so it is hard for me to decide if polars could be good replacement here. 



> Another related suggestion - the [DataFrame API standard](https://github.com/data-apis/dataframe-api) is further along now than it was when we introduced pandas (i.e. there is a draft we might be able to work off now). That would allow napari to be dependency free for dataframe/table data, pushing the choice to the user/plugin-author when they need to support that kind of data (and skipping it when they don't).

We use it as internal storage format. It is not only API case, as we convert to it another formats.

**Comment by DragaDoncila** [comment]
> My two cents are mostly related to what I'm trying to achieve, and specifically just have the image viewer. I want to strip it of the console, the plugin manager, and also the file readers and writers. I would like to just display raw numpy/xarray/dask data and also have access to all the layering mechanism within napari

Fwiw @jacopoabramo that is definitely the end goal for napari-lite. Right now the codebase is not in a state where this is easy to achieve, but this is absolutely what we'll be working towards. 

We definitely understand the pain points with install/launch time and the number of large dependencies, and improving this for the core library is absolutely part of the goal! 

We'll keep updating this issue as we start working on this in the new year

**Comment by psobolewskiPhD** [comment]
Re: the data frame bit, as I mentioned above, I think https://github.com/narwhals-dev/narwhals could potentially be an option. The API is a bit different from pandas, more polars like, but I'm not sure that matters from a user perspective because they'd get native dataframes of whatever type.

**Comment by jacopoabramo** [comment]
> Fwiw @jacopoabramo that is definitely the end goal for napari-lite. Right now the codebase is not in a state where this is easy to achieve, but this is absolutely what we'll be working towards.
> 
> We definitely understand the pain points with install/launch time and the number of large dependencies, and improving this for the core library is absolutely part of the goal!
> 
> We'll keep updating this issue as we start working on this in the new year

@DragaDoncila thanks for the reply; yes I realized that I rushed to adding my comments to the issue while these points have already been cleared well enough; apologies for rushing. Thanks again!
